File: Week_06/621/task_scheduler.py
from typing import List


class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        m = len(tasks)
        if m < 2 or n == 0:
            return m
        counter = [0] * 26

        for t in tasks:
            counter[ord(t) - ord('A')] += 1

        counter.sort()

        max_count = counter[-1]

        res = (max_count - 1) * (n + 1) + 1
        for c in counter[:-1]:
            if c == max_count:
                res += 1
        return max(res, m)

File: Week_06/621/task_scheduler.go
package main

import "sort"

func leastInterval(tasks []byte, n int) int {
	m := len(tasks)
	if m < 2 || n == 0 {
		return m
	}

	counter := make([]int, 26)
	for i := 0; i < m; i++ {
		counter[tasks[i]-'A']++
	}
	sort.Ints(counter)

	maxCount := counter[25]
	x := 0
	for i := 0; i < 25; i++ {
		if counter[i] == maxCount {
			x++
		}
	}
	return max((maxCount-1)*(n+1)+1+x, m)
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}
